salt/transport/ipc.py

```diff
@@ -180,18 +189,52 @@ def __singleton_init__(self, socket_path, io_loop=None):
         self.socket_path = socket_path
         self._closing = False
 
+        self.subscribe_handlers = []
+        self._mid = 1
+        # TODO: move to config
+        self._max_messages = sys.maxint - 1  # number of IDs before we wrap
+
+        # Maximum number of inflight messages
+        self.maxflight = 5  # TODO: config
+
+        # queue of messages to send
+        self.send_queue = []
+
+        # mapping of message_id -> future
+        self.inflight_messages = {}
+
+        # mapping of message_future -> timeout
+        self.timeout_map = {}
+
+    def _alloc_mid(self):
+        '''
+        Return an available MID
+        '''
+        wrap = False
+        while self._mid in self.inflight_messages:
```

Comment:
We don't have to worry about it changing as we iterate over it only because we don't have any `yield` statements within this loop-- which means we never yield execution.